### ISSUE TITLE
Incomplete HTTP request cause 500 Internal Servier Error

### ISSUE THREAD

**Issue by andy-liang** [GrahamDumpleton/mod_wsgi#75]
The follow script cause a 500 response:
# ! /usr/bin/env python

import socket

sock = socket.socket()
sock.connect(("localhost", 80))
sock.send("GET / HTTP/1.1\r\n")
sock.send("Host: localhost\r\n")
sock.send("Content-Length: 10000\r\n")
sock.send("\r\n")
sock.shutdown(socket.SHUT_WR)

print sock.recv(102400)
sock.close()

The httpd error log shows:

[Mon Mar 30 13:43:00 2015] [error] [client 127.0.0.1] mod_wsgi (pid=16857): Request data read error when proxying data to daemon process: End of file found.

**Comment by GrahamDumpleton** [comment]
Which looks entirely reasonable to me.

You sent Content-Length as 10000 but then didn't actually send any request content. The result is that the application when reading the request content found that the advertised content length was bogus as there wasn't that much data. The result was that an exception was raised so that the application knew that reading the request content encountered a problem and the application has not caught that and handled it, resulting in the whole web request failing with a 500 as the exception made it all the way back to mod_wsgi.

Why do you think it should behave any other way? How do you think it should behave when the request content was truncated?

**Comment by andy-liang** [comment]
I think I should clarify what is the application behind mod_wsgi. It looks like this:

def application(_, start_response):
    try:
        start_response("200 OK", [])
        return ["^_^"]
    except:
    	return ["~_~"]

the exception is all internal to mod_wsgi. it should handle the exception and response with 400 Bad Request.

**Comment by andy-liang** [comment]
sorry, looks like GitHub formatted my underscore away. Here is the code again:

```Python
def application(_, start_response):
    try:
        start_response("200 OK", [])
        return ["^_^"]
    except:
    	return ["~_~"]
```

**Comment by GrahamDumpleton** [comment]
Okay, understand why that is the case now. Problem is that I can't readily know specifically why it failed. There could be many reasons and I can't necessarily just assume it was because of a bad request. It need not even be a bad request in this case. The request could have been legitimate but the network connection failed. In most cases there will not even be a connection to send back a response. So a generic 500 response is probably the best that can be hoped for.

**Comment by andy-liang** [comment]
That make sense.

Is there anything I can do to log the wrong Content-Length? The current error seem a bit cryptic.

**Comment by GrahamDumpleton** [comment]
Sorry for slow reply. Catching up after a bad week.

As I mentioned, in this particular case it is a bit hard to know specifically why it failed and thus log a more specific message. It is made harder due to being detected in the Apache child processes where the request is being proxied to daemon processes and where most of the work being done is being done down in Apache code.

Eventually I hope to change how the proxying is done so I have better control over it to allow full duplex HTTP connections with a daemon process and when that is done better logging may be possible, but that is likely to be some time off. It would also probably only be reasonable to log it at debug level as incidences of where this occurred for a HTTP client mistake vs a dropped connection could not be distinguished and logging at error level would simply create lots of noise in the error logs which wouldn't be helpful.

**Comment by bkabrda** [comment]
Hi, I have a similar issue: I have a flask app running on mod_wsgi (version 4.5.7). When a client disconnects prematurely, my flask API endpoint gets executed (and the code returns 400 response), but the server ignores that and returns 500 anyway. I'm not an expert, but there are 2 (possibly 3) ways this should be handled, IMO:
- either my API endpoint shouldn't be touched at all and httpd should return 500 on its own
- or httpd should execute my API endpoint and respect what it returns
- (or make this a configuration value and let users choose which of the 2 above behaviours they want?)
I don't understand why httpd executes my endpoint, but ignores its return value. Maybe I am doing something wrong or misunderstanding something?

**Comment by GrahamDumpleton** [comment]
Closing this as much as can address this, should be addressed by https://github.com/GrahamDumpleton/mod_wsgi/commit/004ae44ae46b410f3999da1a34dd11e469d98169 and will be released in mod_wsgi version 4.5.25. Specifically, where appears that client connection was closed prematurely, the WSGI application HTTP response if already known will take precedence.
